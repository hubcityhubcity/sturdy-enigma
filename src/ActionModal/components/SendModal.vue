<template>
  <ActionModal
    id="send-modal"
    ref="actionModal"
    :validate="validateForm"
    :amount="amount"
    title="Send"
    submission-error-prefix="Sending tokens failed"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    feature-flag="send"
    @close="clear"
    @txIncluded="onSuccess"
  >
    <TmFormGroup
      :error="$v.address.$error && $v.address.$invalid"
      class="action-modal-form-group"
      field-id="send-address"
      field-label="Send To"
    >
      <TmField
        id="send-address"
        v-model.number="$v.address.$model"
        v-focus
        type="text"
        placeholder="Address"
        @keyup.enter.native="refocusOnAmount"
      />
      <TmFormMsg
        v-if="$v.address.$error && !$v.address.required"
        name="Address"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.address.$error && !$v.address.bech32Validate"
        name="Address"
        type="bech32"
      />
    </TmFormGroup>
    <TmFormGroup
      v-if="getDenoms.length > 1"
      :error="$v.selectedToken.$error"
      class="action-modal-form-group"
      field-id="selected-token"
      field-label="Token"
    >
      <TmField
        id="token"
        v-model="selectedToken"
        :title="`Select the token you wish to operate with`"
        :options="getDenoms"
        placeholder="Select the token"
        type="select"
      />
      <TmFormMsg
        v-if="$v.selectedToken.$error && !$v.selectedToken.required"
        name="Token"
        type="required"
      />
    </TmFormGroup>
    <TmFormGroup
      id="form-group-amount"
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span v-if="selectedToken" class="input-suffix max-button">{{
        selectedToken
      }}</span>
      <TmFieldGroup>
        <TmField
          id="amount"
          ref="amount"
          v-model="amount"
          class="tm-field-addon"
          placeholder="0"
          type="number"
          @keyup.enter.native="enterPressed"
        />
        <TmBtn
          type="button"
          class="secondary addon-max"
          value="Set Max"
          @click.native="setMaxAmount()"
        />
      </TmFieldGroup>
      <TmFormMsg
        v-if="selectedBalance.amount === 0"
        :msg="`doesn't have any ${selectedToken}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
      <TmFormMsg
        v-else-if="isMaxAmount()"
        msg="You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
        type="custom"
        class="tm-form-msg--desc max-message"
      />
    </TmFormGroup>
    <a v-if="editMemo === false" id="edit-memo-btn" @click="showMemo()">
      Need to edit the memo field?
    </a>
    <TmFormGroup
      v-else
      id="memo"
      :error="$v.memo.$error && $v.memo.$invalid"
      class="action-modal-group"
      field-id="memo"
      field-label="Memo"
    >
      <TmField
        id="memo"
        v-model="memo"
        v-focus
        type="text"
        @keyup.enter.native="enterPressed"
      />
      <TmFormMsg
        v-if="$v.memo.$error && !$v.memo.maxLength"
        name="Memo"
        type="maxLength"
        :max="max_memo_characters"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import gql from "graphql-tag"
import b32 from "scripts/b32"
import { required, between, decimal, maxLength } from "vuelidate/lib/validators"
import { uatoms, SMALLEST } from "src/scripts/num"
import { mapGetters } from "vuex"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmField from "src/components/common/TmField"
import TmFieldGroup from "src/components/common/TmFieldGroup"
import TmBtn from "src/components/common/TmBtn"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import transaction from "../utils/transactionTypes"
import { toMicroDenom } from "src/scripts/common"
import config from "src/../config"
import { UserTransactionAdded } from "src/gql"

const defaultMemo = "(Sent via Lunie)"

export default {
  name: `send-modal`,
  components: {
    TmField,
    TmFieldGroup,
    TmFormGroup,
    TmFormMsg,
    ActionModal,
    TmBtn
  },
  props: {
    denoms: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    address: ``,
    amount: config.development ? 0.000001 : null, // dev life, hard life > make simple
    memo: defaultMemo,
    max_memo_characters: 256,
    editMemo: false,
    isFirstLoad: true,
    selectedToken: ``,
    balances: []
  }),
  computed: {
    ...mapGetters([`network`]),
    ...mapGetters({ userAddress: `address` }),
    selectedBalance() {
      const selectedBalance = this.balances.filter(
        balance => balance.denom === this.selectedToken || this.denoms[0]
      )
      if (selectedBalance.length > 0) {
        return selectedBalance[0]
      }
      return { amount: 0 }
    },
    transactionData() {
      return {
        type: transaction.SEND,
        toAddress: this.address,
        amounts: [
          {
            amount: uatoms(+this.amount),
            denom: toMicroDenom(this.selectedToken)
          }
        ],
        memo: this.memo
      }
    },
    notifyMessage() {
      return {
        title: `Successful Send`,
        body: `Successfully sent ${+this.amount} ${this.selectedToken}s to ${
          this.address
        }`
      }
    },
    getDenoms() {
      return this.denoms
        ? this.denoms.map(denom => (denom = { key: denom, value: denom }))
        : []
    }
  },
  watch: {
    // we set the amount in the input to zero every time the user selects another token so they
    // realize they are dealing with a different balance each time
    selectedToken: function() {
      if (!this.isFirstLoad) {
        this.amount = 0
      } else {
        this.isFirstLoad = false
      }
    },
    balances: function(balances) {
      if (balances.length === 0) return
      this.selectedToken = balances[0].denom
    }
  },
  mounted() {
    this.$apollo.queries.balances.refetch()
  },
  methods: {
    open() {
      this.$refs.actionModal.open()
    },
    onSuccess(event) {
      this.$emit(`success`, event)
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.address = undefined
      this.amount = undefined
      this.editMemo = false
      this.memo = defaultMemo
      this.sending = false
    },
    setMaxAmount() {
      this.amount = this.selectedBalance.amount
    },
    isMaxAmount() {
      if (this.selectedBalance.amount === 0) {
        return false
      } else {
        return (
          parseFloat(this.amount) === parseFloat(this.selectedBalance.amount)
        )
      }
    },
    token() {
      if (!this.selectedToken) return ``

      return this.selectedToken
    },
    bech32Validate(param) {
      try {
        b32.decode(param)
        return true
      } catch (error) {
        return false
      }
    },
    enterPressed() {
      this.$refs.actionModal.validateChangeStep()
    },
    refocusOnAmount() {
      this.$refs.amount.$el.focus()
    },
    showMemo() {
      this.memo = ``
      this.editMemo = true
    }
  },
  validations() {
    return {
      address: {
        required,
        bech32Validate: this.bech32Validate
      },
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, this.selectedBalance.amount)
      },
      denoms: { required },
      selectedToken: { required },
      memo: {
        maxLength: maxLength(this.max_memo_characters)
      }
    }
  },
  apollo: {
    balances: {
      query: gql`
        query BalancesSendModal($networkId: String!, $address: String!) {
          balances(networkId: $networkId, address: $address) {
            amount
            denom
          }
        }
      `,
      skip() {
        return !this.userAddress
      },
      variables() {
        return {
          networkId: this.network,
          address: this.userAddress
        }
      }
    },
    $subscribe: {
      userTransactionAdded: {
        variables() {
          /* istanbul ignore next */
          return {
            networkId: this.network,
            address: this.userAddress
          }
        },
        skip() {
          /* istanbul ignore next */
          return !this.userAddress
        },
        query: UserTransactionAdded,
        result() {
          /* istanbul ignore next */
          this.$apollo.queries.balances.refetch()
        }
      }
    }
  }
}
</script>
<style scoped>
#edit-memo-btn {
  margin-top: 2.4rem;
  font-size: 12px;
  cursor: pointer;
}

#form-group-amount {
  margin-bottom: 30px;
}
</style>
