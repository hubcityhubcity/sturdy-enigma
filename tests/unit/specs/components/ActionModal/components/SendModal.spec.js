import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import SendModal from "src/ActionModal/components/SendModal"

describe(`SendModal`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.directive(`focus`, () => {})

  let wrapper, $store

  const getters = {
    connected: true,
    session: { signedIn: true, address: "cosmos1234" }
  }

  const state = {}

  beforeEach(async () => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters,
      state
    }

    wrapper = shallowMount(SendModal, {
      localVue,
      mocks: {
        $store,
        $apollo: {
          queries: {
            balance: {
              refetch: () => {}
            }
          }
        }
      },
      propsData: {
        denoms: ["STAKE"]
      },
      sync: false
    })

    wrapper.setData({
      balances: [
        {
          denom: `STAKE`,
          amount: 10000
        }
      ]
    })

    wrapper.vm.$refs.actionModal = {
      submit: cb => cb(),
      open: jest.fn()
    }
    wrapper.vm.open("stake")
  })

  it(`should display send modal form`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`clears on close`, () => {
    const self = {
      $v: { $reset: jest.fn() },
      address: `cosmos1address`,
      amount: 10
    }
    SendModal.methods.clear.call(self)
    expect(self.$v.$reset).toHaveBeenCalled()
    expect(self.address).toBe(undefined)
    expect(self.amount).toBe(undefined)
  })

  it(`shows the memo input if desired`, async () => {
    wrapper.find("#edit-memo-btn").trigger("click")
    await wrapper.vm.$nextTick()

    expect(wrapper.find("#memo").isVisible()).toBe(true)
  })

  describe(`validation`, () => {
    it(`should show address required error`, async () => {
      wrapper.setProps({
        denom: `STAKE`
      })
      wrapper.setData({
        address: ``,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })
    it(`should show bech32 error when address length is too short`, async () => {
      wrapper.setProps({
        denom: `STAKE`
      })
      wrapper.setData({
        address: `asdf`,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })

    it(`should show bech32 error when address length is too long`, async () => {
      wrapper.setProps({
        denom: `STAKE`
      })
      wrapper.setData({
        address: `asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf`,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })
    it(`should show bech32 error when alphanumeric is wrong`, async () => {
      wrapper.setData({
        address: ``
      })
      expect(wrapper.vm.validateForm()).toBe(false)
      await wrapper.vm.$nextTick()
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  it(`should submit when enterPressed is called`, async () => {
    const self = {
      $refs: { actionModal: { validateChangeStep: jest.fn() } }
    }
    SendModal.methods.enterPressed.call(self)
    expect(self.$refs.actionModal.validateChangeStep).toHaveBeenCalled()
  })

  it(`should refocus on amount when focusOnAmount is called`, async () => {
    const self = {
      $refs: { amount: { $el: { focus: jest.fn() } } }
    }
    SendModal.methods.refocusOnAmount.call(self)
    expect(self.$refs.amount.$el.focus).toHaveBeenCalled()
  })

  it(`validates bech32 addresses`, () => {
    expect(
      wrapper.vm.bech32Validate(`cosmos1x7wzdumfj8pncd99mqc0mkqfrrps3l3pjz8tk6`)
    ).toBe(true)
    expect(
      wrapper.vm.bech32Validate(`cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`)
    ).toBe(false)
  })

  it("should return transaction data in correct form", () => {
    wrapper.setProps({
      denom: `STAKE`
    })
    wrapper.setData({
      address: `cosmos12345`,
      amount: 2
    })
    expect(wrapper.vm.transactionData).toEqual({
      type: "MsgSend",
      toAddress: "cosmos12345",
      amounts: [{ amount: "2000000", denom: "stake" }],
      memo: "(Sent via Lunie)"
    })
  })

  it("should return notification message", () => {
    wrapper.setProps({
      denom: `STAKE`
    })
    wrapper.setData({
      address: `cosmos12345`,
      amount: 2
    })
    expect(wrapper.vm.notifyMessage).toEqual({
      title: `Successful Send`,
      body: `Successfully sent 2 STAKEs to cosmos12345`
    })
  })

  describe(`if amount field max button clicked`, () => {
    it(`amount has to be 10000 atom`, async () => {
      wrapper.setProps({
        denom: `STAKE`
      })
      wrapper.setData({
        address: `cosmos12345`
      })
      wrapper.vm.setMaxAmount()
      expect(wrapper.vm.amount).toBe(10000)
    })
    it(`should show warning message`, async () => {
      wrapper.setProps({
        denom: `STAKE`
      })
      wrapper.setData({
        address: `cosmos12345`
      })
      wrapper.vm.setMaxAmount()
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).toContain(
        "You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
      )
    })
    it(`should not show warning message if balance = 0`, async () => {
      wrapper.setData({
        balances: [
          {
            amount: 0,
            denom: "STAKE"
          }
        ]
      })
      wrapper.vm.setMaxAmount()
      await wrapper.vm.$nextTick()
      expect(wrapper.html()).not.toContain(
        "You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
      )
    })
    it(`isMaxAmount() should return false if balance = 0`, async () => {
      wrapper.setData({
        balances: [
          {
            amount: 0,
            denom: "STAKE"
          }
        ]
      })
      wrapper.vm.setMaxAmount()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.isMaxAmount()).toBe(false)
    })
  })

  describe(`Set token and balance`, () => {
    it(`it takes the corresponding balance from the balances array when
    selectedToken has been chosen and balances has a length over 1`, async () => {
      wrapper.setData({
        selectedToken: `TOKEN1`,
        balances: [
          {
            amount: 1,
            denom: "TOKEN1"
          },
          {
            amount: 2,
            denom: "TOKEN2"
          }
        ]
      })
      wrapper.vm.setTokenAndBalance()
      expect(wrapper.vm.selectedBalance.amount).toBe(1)
    })
    // This one creates a lot of ugly errors
    // it(`returns empty string if selectedToken hasn't been chosen yet`, () => {
    //   wrapper.setData({
    //     balances: []
    //   })
    //   const res = wrapper.vm.token()
    //   expect(res).toBe("")
    // })
    it(`returns selectedToken if selectedToken has been chosen`, () => {
      const res = wrapper.vm.token()
      expect(res).toBe("STAKE")
    })
  })
})
