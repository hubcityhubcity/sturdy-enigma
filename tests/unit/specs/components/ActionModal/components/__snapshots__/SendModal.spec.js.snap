// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendModal should display send modal form 1`] = `
<actionmodal-stub
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address"
      type="text"
      value=""
    />
     
    <!---->
  </tmformgroup-stub>
   
  <!---->
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <span
      class="input-suffix max-button"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <a
    id="edit-memo-btn"
  >
    
    Need to edit the memo field?
  
  </a>
</actionmodal-stub>
`;

exports[`SendModal validation should show address required error 1`] = `
<actionmodal-stub
  amount="2"
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address"
      type="text"
      value=""
    />
     
    <tmformmsg-stub
      msg=""
      name="Address"
      type="required"
    />
  </tmformgroup-stub>
   
  <!---->
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <span
      class="input-suffix max-button"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="2"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <a
    id="edit-memo-btn"
  >
    
    Need to edit the memo field?
  
  </a>
</actionmodal-stub>
`;

exports[`SendModal validation should show bech32 error when address length is too long 1`] = `
<actionmodal-stub
  amount="2"
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address"
      type="text"
      value="asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf"
    />
     
    <tmformmsg-stub
      msg=""
      name="Address"
      type="bech32"
    />
  </tmformgroup-stub>
   
  <!---->
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <span
      class="input-suffix max-button"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="2"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <a
    id="edit-memo-btn"
  >
    
    Need to edit the memo field?
  
  </a>
</actionmodal-stub>
`;

exports[`SendModal validation should show bech32 error when address length is too short 1`] = `
<actionmodal-stub
  amount="2"
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address"
      type="text"
      value="asdf"
    />
     
    <tmformmsg-stub
      msg=""
      name="Address"
      type="bech32"
    />
  </tmformgroup-stub>
   
  <!---->
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <span
      class="input-suffix max-button"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="2"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <!---->
  </tmformgroup-stub>
   
  <a
    id="edit-memo-btn"
  >
    
    Need to edit the memo field?
  
  </a>
</actionmodal-stub>
`;

exports[`SendModal validation should show bech32 error when alphanumeric is wrong 1`] = `
<actionmodal-stub
  featureflag="send"
  id="send-modal"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Sending tokens failed"
  title="Send"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="send-address"
    fieldlabel="Send To"
  >
    <tmfield-stub
      id="send-address"
      placeholder="Address"
      type="text"
      value=""
    />
     
    <tmformmsg-stub
      msg=""
      name="Address"
      type="required"
    />
  </tmformgroup-stub>
   
  <!---->
   
  <tmformgroup-stub
    class="action-modal-form-group"
    error="true"
    fieldid="amount"
    fieldlabel="Amount"
    id="form-group-amount"
  >
    <span
      class="input-suffix max-button"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <tmformmsg-stub
      msg=""
      name="Amount"
      type="required"
    />
  </tmformgroup-stub>
   
  <a
    id="edit-memo-btn"
  >
    
    Need to edit the memo field?
  
  </a>
</actionmodal-stub>
`;
